Broadcast transactions only after their timelock is up

TheBlueMatt opened the pull request:
Built on #928, this avoids broadcasting transactions too early resulting in us leaning on users to do rebroadcasting. Instead, we wait until timelocks are up to broadcast.

ariard commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -1903,6 +1905,11 @@ impl<Signer: Sign> ChannelMonitorImpl<Signer> {
+					continue;
 				} else { None };
 				if let Some(htlc_tx) = self.onchain_tx_handler.get_fully_signed_htlc_tx(
 					&::bitcoin::OutPoint { txid, vout }, &preimage) {
```
nit: you can rename `res` to `htlc_success_txn`

ariard commented on lightning/src/chain/onchaintx.rs:
```diff
@@ -371,24 +389,28 @@ impl<ChannelSigner: Sign> OnchainTxHandler<ChannelSigner> {
+			} else {
+				for locked_package in self.locktimed_packages.iter().map(|v| v.1.iter()).flatten() {
+					if locked_package.outpoints() == req.outpoints() {
+						continue 'req_walk;
```
Duplicate the above logger, as it's already registered, maybe mentioning they're registered in the special locktime buffer.

ariard commented on lightning/src/chain/onchaintx.rs:
```diff
@@ -404,10 +426,26 @@ impl<ChannelSigner: Sign> OnchainTxHandler<ChannelSigner> {
+				for outpoint in req.outpoints() {
+					log_debug!(logger, "  Outpoint {}", outpoint);
+				}
+				self.locktimed_packages.entry(req.package_timelock()).or_insert(Vec::new()).push(req);
```
I think I would gather all preprocessing checks (aggregation, delayed-claiming) in the same for loop above and the restoring loop as the first thing in `update_claims_view`. Post-anchor, we might be interested to aggregate our HTLC-timeout with an equal cltv_expiry and it won't work with this code organization.

ariard commented on lightning/src/chain/onchaintx.rs:
```diff
@@ -404,10 +426,26 @@ impl<ChannelSigner: Sign> OnchainTxHandler<ChannelSigner> {
+					log_debug!(logger, "  Outpoint {}", outpoint);
+				}
+				self.locktimed_packages.entry(req.package_timelock()).or_insert(Vec::new()).push(req);
+			} else if let Some((new_timer, new_feerate, tx)) = self.generate_claim_tx(height, &req, &*fee_estimator, &*logger) {
```
Can we broadcast delayed-transactions scenario with the following reorg-scenario ?
* delayed-transactions are cached in `locktimed_packages`
* height expiration is reached, we generate and broadcast delayed transactions
* chain is reorg-out, delayed transactions are evicted from mempools
* before syncing to new tip, we receive a preimage offchain and enter into `update_claims_view`

I don't think we can regenerate and broadcast a non-propagating delayed transaction as regeneration timer are height-based. In the future, if we switch regeneration timer to mempool feeration variations, we might broadcast such delayed transactions.

Otherwise, I can't think of other edgy reorg cases.

TheBlueMatt replied:
In that case don't we still generate the relevant commitment + timeout/claim transactions when the block comes back in when we sync to the new tip? Either way we should bump after a few blocks if we haven't confirmed by then.

ariard replied:
Clearly, not a big deal. It's a just a edge-case about new feature added by this PR in case of a hypothetical future change. If we switch or add the option of mempool-based timing we'll have to split our hairs on few others things around OnchainTxHandler.

(Thought I still think mempool fluctuations-based timing might be more more interesting for fee savings than dumb rebroadcast after blocks X, if you're not near timelocks expiration)

ariard commented on lightning/src/chain/package.rs:
```diff
@@ -576,13 +600,22 @@ impl PackageTemplate {
 		self.height_timer = cmp::min(self.height_timer, merge_from.height_timer);
 	}
-	pub(crate) fn package_amount(&self) -> u64 {
+	/// Gets the amount of all outptus being spent by this package, only valid for malleable
```
nit: *outputs an "only valid for now (legacy channel types)"

TheBlueMatt replied:
Hmm, not sure what you're suggesting? I'll let you update the docs for anchor in the anchor PRs.

ariard commented on lightning/src/chain/package.rs:
```diff
@@ -374,6 +387,17 @@ impl PackageSolvingData {
 		}
 	}
+	fn absolute_tx_timelock(&self, input_height: u32) -> u32 {
+		let amt = match self {
```
I think input_height should be called "output_conf_height". I concede that height original isn't really clear, but it avoids at least letting think that this timelock are only CSV ones.

Maybe rename `amt` to `absolute_height`?

And precise the +1 is to follow mempool rules of evaluating timelocks, otherwise a newcomer can be easily confused on this.

ariard commented on lightning/src/chain/package.rs:
```diff
@@ -576,13 +600,22 @@ impl PackageTemplate {
 		amounts
 	}
+	pub(crate) fn package_timelock(&self) -> u32 {
+		let mut max_locktime = 0;
```
nit: min_locktime? I would rather call the maximum locktime `std::u32::MAX`

TheBlueMatt replied:
Hmm? This is calculating the maximum locktime of all included inputs, so its still a maximum?

ariard replied:
Ah gotcha, I was thinking about the max absolute value that `nLocktime` might be. `max_default_locktime` would be a bit clearer.

TheBlueMatt replied:
I switched to an iterator as Jeff suggested below, so there is no variable name to be confused about :).

TheBlueMatt commented:
Rebased after merge of #928 and addressed comments.

ariard reviewed (commented):
Code Review ACK 2f755b1

I went through all the fixes commit, globally satisfied with new comments. I had a look on test changes they look good overall.

jkczyz commented on lightning/src/ln/functional_test_utils.rs:
```diff
@@ -212,15 +212,15 @@ pub struct Node<'a, 'b: 'a, 'c: 'b> {
 	pub network_chan_count: Rc<RefCell<u32>>,
 	pub logger: &'c test_utils::TestLogger,
-	pub blocks: RefCell<Vec<(BlockHeader, u32)>>,
+	pub blocks: Arc<Mutex<Vec<(BlockHeader, u32)>>>,
```
Why does this need to be an `Arc`?

TheBlueMatt replied:
Ah, see https://github.com/rust-bitcoin/rust-lightning/pull/932#discussion_r641619700, but both the `Node` and the `TestBroadcaster` hold a reference to the same set of blocks.

TheBlueMatt replied:
Right, but the objects which own the broadcaster are thread-safe, so we have to be :/

jkczyz commented on lightning/src/util/test_utils.rs:
```diff
@@ -201,6 +202,7 @@ impl<Signer: keysinterface::Sign> channelmonitor::Persist<Signer> for TestPersis
 
 pub struct TestBroadcaster {
 	pub txn_broadcasted: Mutex<Vec<Transaction>>,
+	pub blocks: Arc<Mutex<Vec<(BlockHeader, u32)>>>,
```
Likewise, why is an `Arc` needed?

TheBlueMatt replied:
Hmm, not sure what the "Likewise" here is, but `TestBroadcaster` gets passed into various structs with a reference to the same blocks as the `Node` object holds. It failed to compile without being thread-safe (ie arc+mutex), but I didn't investigate further as it seemed sensible that it wouldn't work with an Rc.

jkczyz commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -8689,6 +8700,9 @@ fn test_concurrent_monitor_claim() {
 	let header = BlockHeader { version: 0x20000000, prev_blockhash: Default::default(), merkle_root: Default::default(), time: 42, bits: 42, nonce: 42 };
+	// Make the tx_broadcaster aware of enough blocks that it doesn't think we're violating
+	// transaction lock time requirements here.
+	chanmon_cfgs[0].tx_broadcaster.blocks.lock().unwrap().resize((CHAN_CONFIRM_DEPTH + 1 + TEST_FINAL_CLTV + LATENCY_GRACE_PERIOD_BLOCKS) as usize, (header, 0));
```
Is there a need to use a `Vec`? Would storing the last block header and height suffice?

jkczyz reviewed (commented):
Need to take a closer look at the tests in the later commits.

TheBlueMatt commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -1499,16 +1499,17 @@ fn test_duplicate_htlc_different_direction_onchain() {
 
 	// Check we only broadcast 1 timeout tx
 	let claim_txn = nodes[0].tx_broadcaster.txn_broadcasted.lock().unwrap().clone();
-	let htlc_pair = if claim_txn[0].output[0].value == 800_000 / 1000 { (claim_txn[0].clone(), claim_txn[1].clone()) } else { (claim_txn[1].clone(), claim_txn[0].clone()) };
```
Entirely, the `800_000` wasn't even right anymore - now that we subtract fees no output was ever exactly the 800k msat

TheBlueMatt commented on lightning/src/chain/package.rs:
```diff
@@ -274,8 +274,8 @@ impl PackageSolvingData {
 			// should change with the remaining anchor output patchset.
-			PackageSolvingData::HolderHTLCOutput(..) => { 0 },
-			PackageSolvingData::HolderFundingOutput(..) => { 0 },
+			PackageSolvingData::HolderHTLCOutput(..) => { debug_assert!(false); 0 },
+			PackageSolvingData::HolderFundingOutput(..) => { debug_assert!(false); 0 },
```
Actually, yea. I was avoiding doing it to keep the panic in debug, but I think you're right, if we ever *actually* get there, its probably indication that we're gonna generate an invalid spend transaction and/or burn funds to fee, which is not ok.

TheBlueMatt commented on lightning/src/chain/onchaintx.rs:
```diff
@@ -376,9 +402,17 @@ impl<ChannelSigner: Sign> OnchainTxHandler<ChannelSigner> {
+					if locked_package.outpoints() == req.outpoints() {
+						continue 'req_walk;
+					}
+				}
```
Hmm, I'm not really sure this is at all more readable, but maybe I just fine long chains of iterators a bit harder to read. Thoughts?

```
				let locked_package_timelock = self.locktimed_packages.iter().map(|v| v.1.iter()).flatten()
					.find_map(|locked_package| if locked_package.outpoints() == req.outpoints() {
						Some(locked_package.package_timelock())
					} else { None });
				if let Some(timelock) = locked_package_timelock {
					log_trace!(logger, "Ignoring second claim for outpoint {}:{}, we already have one which we're waiting on a timelock at {} for.",
						req.outpoints()[0].txid, req.outpoints()[0].vout, timelock);
					continue;
				}
```

jkczyz replied:
Given `find` returns an `Option`, you can use that instead of `find_map` with one line:

```rust
.find(|locked_package| locked_package.outpoints() == req.outpoints())
```

and keep the `log_trace` as it was.

TheBlueMatt replied:
Oh! I misread the docs for `find`. Took your suggestion.

TheBlueMatt commented on lightning/src/chain/onchaintx.rs:
```diff
@@ -394,10 +428,26 @@ impl<ChannelSigner: Sign> OnchainTxHandler<ChannelSigner> {
 			preprocessed_requests.push(req);
 		}
 
+		loop {
+			let pop_height = if let Some(first_entry) = self.locktimed_packages.iter().next() {
+				if *first_entry.0 <= height + 1 {
+					*first_entry.0
+				} else { break; }
+			} else { break; };
+			log_trace!(logger, "Restoring delayed claim of package(s) at their timelock at {}.", pop_height);
+			preprocessed_requests.append(&mut self.locktimed_packages.remove(&pop_height).unwrap());
+		}
```
No, I think you're very right - I was thking that because we don't have a `drain` we shouldn't as we'll end up copying a tone, but it doesn't actually matter because we can `drain` (or `append`-from) the inner `Vec`. It does make it much more readable.

TheBlueMatt commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -1907,9 +1907,11 @@ impl<Signer: Sign> ChannelMonitorImpl<Signer> {
+					// Don't broadcast HTLC-Timeout transactions immediately as they don't meet the
+					// current locktime requirements on-chain. We will broadcast them in
+					// `block_confirmed` when `would_broadcast_at_height` returns true.
+					// Note that we add + 1 as transactions are broadcastable when they can be
+					// confirmed in the next block, not the current height.
```
Right, its just noting that we can always broadcast when the locktime is `X + 1`. I can drop if you think its obvious but Antoine suggested it.

* Fun fact of the day - Bitcoin Core had a bug for many years where it actually wouldn't accept-for-relay transactions until they were confirm-able in the *current* block, not X+1.

TheBlueMatt replied:
Right, if the current height is X, then the next block which will be mined is X + 1. Thus, we can always broadcast something with a CLTV expiry of X + 1 as it can go into the next block (but could not appear in the current best block).

jkczyz replied:
Ah, that makes sense. I think dropping "not the current height" would be beneficial as including it can lead to confusion.

jkczyz commented on lightning/src/ln/reorg_tests.rs:
```diff
@@ -102,16 +102,19 @@ fn do_test_onchain_htlc_reorg(local_commitment: bool, claim: bool) {
 		header.prev_blockhash = nodes[1].best_block_hash();
-		connect_block(&nodes[1], &Block { header, txdata: vec![node_2_commitment_txn[0].clone(), node_1_commitment_txn[0].clone()] });
+		let block = Block { header, txdata: vec![node_2_commitment_txn[0].clone(), node_1_commitment_txn[2].clone()] };
+		std::mem::drop(node_1_commitment_txn);
```
Consider just cloning the entire `Vec` rather than holding a lock, given you're gonna clone some elements anyway.

jkczyz commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -5854,6 +5840,13 @@ fn test_key_derivation_params() {
 	assert_ne!(chan_0.3.output[0].script_pubkey, chan_1.3.output[0].script_pubkey);
 
+	// Ensure all nodes are at the same height
+	let node_max_height = std::cmp::max(nodes[0].blocks.lock().unwrap().len(),
+		std::cmp::max(nodes[1].blocks.lock().unwrap().len(), nodes[2].blocks.lock().unwrap().len())) as u32;
```
Could use `max` on an iterator here, too.

jkczyz reviewed (commented):
Just a couple nits, but otherwise this is still outstanding: https://github.com/rust-bitcoin/rust-lightning/pull/932/commits/ff8ca028d538ec1e695aba814eb8be557a20544d#r641682315

TheBlueMatt commented:
Hmm, your link isn't working for me now, but I did try to respond at https://github.com/rust-bitcoin/rust-lightning/pull/932/commits/ff8ca028d538ec1e695aba814eb8be557a20544d#r641820784

ariard reviewed (commented):
Code Review ACK a59b879

Changes since last time are addressing Jeff's comments

TheBlueMatt commented:
Tweaked the comment as suggested at https://github.com/rust-bitcoin/rust-lightning/pull/932#discussion_r641829233, fixed one trivial error from new fixes, and squashed. Will merge after CI:

```
$ git diff-tree -U1 a59b8798 d51d606ba
diff --git a/lightning/src/chain/channelmonitor.rs b/lightning/src/chain/channelmonitor.rs
index 3df0111f5..26187ce34 100644
--- a/lightning/src/chain/channelmonitor.rs
+++ b/lightning/src/chain/channelmonitor.rs
@@ -1914,3 +1914,3 @@ impl<Signer: Sign> ChannelMonitorImpl<Signer> {
 					// Note that we add + 1 as transactions are broadcastable when they can be
-					// confirmed in the next block, not the current height.
+					// confirmed in the next block.
 					continue;
diff --git a/lightning/src/ln/functional_tests.rs b/lightning/src/ln/functional_tests.rs
index 30b9d0145..eb4f6054b 100644
--- a/lightning/src/ln/functional_tests.rs
+++ b/lightning/src/ln/functional_tests.rs
@@ -5843,3 +5843,3 @@ fn test_key_derivation_params() {
 	// Ensure all nodes are at the same height
-	let node_max_height = nodes.iter().map(|node| node.blocks.borrow().len()).max().unwrap() as u32;
+	let node_max_height = nodes.iter().map(|node| node.blocks.lock().unwrap().len()).max().unwrap() as u32;
 	connect_blocks(&nodes[0], node_max_height - nodes[0].best_block_info().1);
$
```
